### PR TITLE
DPL-202 Subscribe to messages

### PR DESCRIPTION
**Pull request by sdjmchattie**
Part of #469 

Changes proposed in this pull request:

* Set up a background consumer for messages on the CRUD queue.
* Add all credentials required to consume messages.
* Provide a setup script for local development to ensure vhost and queue is installed on the local RabbitMQ instance.
* Update the README with instructions for the above.

* Consumes message continuously from the CRUD queue.  At the moment logging them to the console.  Next update this will be a callback method so that we can process them into the database and create a feedback message.

### CODE REVIEW

**Comment by sdjmchattie** [comment]
We should discuss what level of coverage we want in unit tests for these new files.  They may only add limited value since the code was provided by Pika and then adapted to our needs.

File: crawler/__init__.py [hunk 1]
```diff
@@ -44,3 +47,21 @@ def setup_routes(app):
...
         app.register_blueprint(v1_routes.bp, url_prefix="/v1")
+
+
+def start_rabbit_consumer(app):
+    # Flask in debug mode spawns a child process so that it can restart the process each time your code changes,
+    # the new child process initializes and starts a new consumer causing more than one to exist.
+    if flask.helpers.get_debug_flag() and not werkzeug.serving.is_running_from_reloader():
+        return
+
+    use_ssl = app.config["RABBITMQ_SSL"]
```
Comment (stevieing) [inline crawler/__init__.py:58]: It almost feels like an object would be good here ...

**Review by stevieing — APPROVED** [review]
Is it possible to turn this into a library that other teams could use e.g. ToL?

**Comment by sdjmchattie** [comment]
> Is it possible to turn this into a library that other teams could use e.g. ToL?

I think this is a nice idea @stevieing and certainly worthwhile coming back to later.  For now though, with the complications that are already present, I think it would be best to keep this as is and think about extracting it to a library later if that would be useful.

**Comment by stevieing** [comment]
I was not asking for it to be done as part of this work :).

More an enquiry as to whether it could be done. I think it would be really good to consider when we pitch using a message queue for ToL work requests.

On 5 Apr 2022, at 17:18, Stuart McHattie ***@***.******@***.***>> wrote:



Is it possible to turn this into a library that other teams could use e.g. ToL?

I think this is a nice idea @stevieing [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_stevieing&d=DwMCaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=PioAYl_Zz9lnGwJRms8MsEB0E-grNRW-ut-_gAGl70g&m=3WZ-ll0UepegA5FWiJdae5u0Hq8CqcWEaFmp7tRjHT_xH9tFAF4wE2eC0nicMgTv&s=f5mEziM9sU7Kiy2FASAwhcW2e8Zfpg33Gs59mNxJy7E&e=> and certainly worthwhile coming back to later. For now though, with the complications that are already present, I think it would be best to keep this as is and think about extracting it to a library later if that would be useful.

—
Reply to this email directly, view it on GitHub [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sanger_crawler_pull_544-23issuecomment-2D1088959730&d=DwMCaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=PioAYl_Zz9lnGwJRms8MsEB0E-grNRW-ut-_gAGl70g&m=3WZ-ll0UepegA5FWiJdae5u0Hq8CqcWEaFmp7tRjHT_xH9tFAF4wE2eC0nicMgTv&s=AO4igWyGPB8NiHFMAFy_PFtZ_ZnyopO8y8U8XYs_8qM&e=>, or unsubscribe [github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAD6ZH2INGS57NKFT2HLPN3VDRRW7ANCNFSM5SP7LFPA&d=DwMCaQ&c=D7ByGjS34AllFgecYw0iC6Zq7qlm8uclZFI0SqQnqBo&r=PioAYl_Zz9lnGwJRms8MsEB0E-grNRW-ut-_gAGl70g&m=3WZ-ll0UepegA5FWiJdae5u0Hq8CqcWEaFmp7tRjHT_xH9tFAF4wE2eC0nicMgTv&s=4RwLZU_K9T2ELLNe2cH-YQBbFAuBY_ICAQ6yPVvYBMo&e=>.
You are receiving this because you were mentioned.Message ID: ***@***.***>




-- 
 The Wellcome Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Review by stevieing — APPROVED** [review]

**Pull request merged**